### ISSUE TITLE
Improve trace callback performance

### ISSUE THREAD

**Issue by pschanely** [pschanely/CrossHair#115]
[As described here](https://github.com/pschanely/CrossHair/discussions/105#discussioncomment-662381), CrossHair spends a lot of its execution time in a pure-python `sys.settrace` handler. Making this faster should speed up CrossHair quite a bit.

It would be great to benchmark how much would be saved by moving some of that logic to an extension module written in C; it wouldn't be possible to move all the logic, but many tracing invocations are no-ops, and avoiding those at least could help a lot. [Cython](https://cython.org/) might also be worth investigating.

**Comment by pschanely** [comment]
Much of this is now implemented, but there are two more valuable changes:
(1) We should port [the many different ways to do a stack read and write](https://github.com/pschanely/CrossHair/blob/9d8bceb4ff8d93afc644a7ad98a6a956493ec07d/crosshair/tracers.py#L33) to C. In theory, the resulting code should be faster, smaller, and more correct/portable, because the `ctypes` approach we have now is overly dependent on the byte-level layout of the frame struct.
(2) [CompositeTracer](https://github.com/pschanely/CrossHair/blob/9d8bceb4ff8d93afc644a7ad98a6a956493ec07d/crosshair/tracers.py#L295) is now largely just a wrapper around [CTracer](https://github.com/pschanely/CrossHair/blob/9d8bceb4ff8d93afc644a7ad98a6a956493ec07d/crosshair/_tracers.c#L453). Let's try to remove the CompositeTracer layer entirely.

**Comment by aliviahossain** [comment]
Hi! I'm interested in contributing to this issue. I found this issue on Issuehunt. While I'm still learning about performance optimization and working with C/Cython in Python, I’d love to give this a try and learn through the process.

Would it be okay if I took this on? I’ll make sure to ask questions and share progress as I go.

Thanks!

**Comment by nyashahama** [comment]
Hi @pschanely, I’m interested in helping with the remaining work for this IssueHunt bounty and wanted to check in first so I don’t duplicate the work in #409.

Based on the latest maintainer guidance, my understanding is that item (2) may already be covered by #409, and that the most valuable remaining path is likely item (3): porting the frequently executed `TracingModule` / `trace_op` logic into C, including the related constants and `handle_call_*` paths.

Could you confirm whether that’s still the preferred scope for closing this issue? If so, I’ll start by putting together a short design/profiling note covering the current hot path, the proposed C boundary, packaging impact, and test plan before opening an implementation PR.

**Comment by pschanely** [comment]
> Hi [@pschanely](https://github.com/pschanely), I’m interested in helping with the remaining work for this IssueHunt bounty and wanted to check in first so I don’t duplicate the work in [#409](https://github.com/pschanely/CrossHair/pull/409).
> 
> Based on the latest maintainer guidance, my understanding is that item (2) may already be covered by [#409](https://github.com/pschanely/CrossHair/pull/409), and that the most valuable remaining path is likely item (3): porting the frequently executed `TracingModule` / `trace_op` logic into C, including the related constants and `handle_call_*` paths.
> 
> Could you confirm whether that’s still the preferred scope for closing this issue? If so, I’ll start by putting together a short design/profiling note covering the current hot path, the proposed C boundary, packaging impact, and test plan before opening an implementation PR.

Greetings! You have the right idea about what's remaining. Take a look at #410 also, which deals with the handle_call_* pieces. I guess with the advent of our new AI friends, there has been a lot of interest in this issue lately. So, please be aware that I can't guarantee that your otherwise helpful changes don't go to waste!

**Comment by evilfps** [comment]
Thanks for clarifying the remaining scope.

I am a student and this bounty would help, but I do not want to duplicate work already covered by #410 or #411. If there is still a specific TracingModule or trace_op subtask you would prefer someone to take, I can work on that and send a small tested PR.

If the current PRs are already likely to cover it, I will wait instead.

**Comment by Rajesh270712** [comment]
Hi @pschanely, I re-read #115 after #410 and #413 landed. I would avoid the merged call-stack decoding and target-normalization paths.

If item (3) is still worth another small slice, the residual area I see is the remaining `TracingModule.trace_op` dispatch path around the Python callback boundary. I would start with a short profiling note against current `main`, then keep any implementation limited to the tracer C/Python boundary: preserve `trace_call()` as the extension hook, keep `Untracable` and keyword-argument behavior unchanged, and only consider moving no-op/dispatch wrapper work into C if the profile supports it.

Validation would include rebuilding the extension and running the focused tracer tests plus representative core/builtins tests before opening any PR.

If you would rather review #409/#394 first, or if #410/#413 have covered enough of item (3), I can stand down. If this residual slice would still be useful, could you confirm whether a small PR from @Rajesh270712 on that scope is welcome?

**Comment by pschanely** [comment]
> If you would rather review [#409](https://github.com/pschanely/CrossHair/pull/409)/[#394](https://github.com/pschanely/CrossHair/pull/394) first, or if [#410](https://github.com/pschanely/CrossHair/pull/410)/[#413](https://github.com/pschanely/CrossHair/pull/413) have covered enough of item (3), I can stand down. If this residual slice would still be useful, could you confirm whether a small PR from [@Rajesh270712](https://github.com/Rajesh270712) on that scope is welcome?

Go for it! The PRs you referenced haven't had activity in a while, so it seems unlikely that they'll get over the finish line.

**Comment by Rajesh270712** [comment]
Hi, I opened a focused PR for this: https://github.com/pschanely/CrossHair/pull/435. It keeps the change to the residual `TracingModule.trace_op` / `trace_call` dispatch path: trace-call-only modules can use the C tracer fast path, while custom `trace_op`, custom `__call__`, keyword-expansion, and `Untracable` cases stay on the existing behavior. I validated it with the tracer, enforcement/opcode, core/path/diff, and smoke test groups listed in the PR. If this approach still looks aligned, could you assign the issue to @Rajesh270712? Happy to adjust if you prefer a different structure.

**Comment by pschanely** [comment]
I think this is now complete, or close enough to call it that!
I'm not super familiar with how issuehunt bounties work, but I believe that after I close this, I can assign contributor ratios. Thanks everyone!

**Comment by pschanely** [comment]
I'd intended this is get evenly distributed among the merged PRs, but either (1) issuehunt doesn't actually support this, or (2) I didn't use it correctly. Sent an email out to try and sort it out, but not confident anything can be done.

Contributors: please accept my sincere gratitude (and in many cases, apologies).
And, @andrew222651 - thank you so much for funding - in total, this work has been incredible for CrossHair effectiveness!
